JitsiConference.js

```diff
@@ -1623,7 +1795,519 @@ JitsiConference.prototype.broadcastEndpointMessage = function(payload) {
+                },
+                error => {
+                    logger.error(
+                        'Failed to accept incoming P2P Jingle session', error);
+                }
+            );
+        },
+        error => {
+            logger.error(
+                `Failed to add ${localTracks} to the P2P connection`, error);
+        });
+};
+
+/**
+ * Attaches local tracks back to the JVB connection.
+ * @private
+ */
+JitsiConference.prototype._attachLocalTracksToJvbSession = function() {
+    const localTracks = this.getLocalTracks();
+
+    logger.info(`Attaching ${localTracks} to JVB`);
+    this.jvbJingleSession.attachLocalTracks(localTracks).then(
+        () => {
+            logger.info(`Attach ${localTracks} to JVB success!`);
+        },
+        error => {
+            logger.error(`Attach ${localTracks} to JVB failed!`, error);
+        });
+};
+
+/**
+ * Adds remote tracks to the conference associated with the JVB session.
+ * @private
+ */
+JitsiConference.prototype._addRemoteJVBTracks = function() {
+    this._addRemoteTracks('JVB', this.jvbJingleSession);
+};
+
+/**
+ * Adds remote tracks to the conference associated with the P2P session.
+ * @private
+ */
+JitsiConference.prototype._addRemoteP2PTracks = function() {
+    this._addRemoteTracks('P2P', this.p2pJingleSession);
+};
+
+/**
+ * Generates fake "remote track added" events for given Jingle session.
+ * @param {string} logName the session's nickname which will appear in log
+ * messages.
+ * @param {JingleSessionPC} jingleSession the session for which remote tracks
+ * will be added.
+ * @private
+ */
+JitsiConference.prototype._addRemoteTracks = function(logName, jingleSession) {
+    if (!jingleSession) {
+        logger.info(`Not adding remote ${logName} tracks - no session yet`);
+
+        return;
+    }
+    const remoteTracks = jingleSession.peerconnection.getRemoteTracks();
+
+    for (const track of remoteTracks) {
+        logger.info(`Adding remote ${logName} track: ${track}`);
+        this.rtc.eventEmitter.emit(RTCEvents.REMOTE_TRACK_ADDED, track);
+    }
+};
+
+/**
+ * Called when {@link XMPPEvents.CONNECTION_ESTABLISHED} event is
+ * triggered for a {@link JingleSessionPC}. Switches the conference to use
+ * the P2P connection if the event comes from the P2P session.
+ * @param {JingleSessionPC} jingleSession the session instance.
+ * @private
+ */
+JitsiConference.prototype._onP2PConnectionEstablished
+= function(jingleSession) {
+    if (!jingleSession.isP2P) {
+
+        return;
+    } else if (this.p2pJingleSession !== jingleSession) {
+        logger.error('CONNECTION_ESTABLISHED - not P2P session ?!');
+
+        return;
+    }
+
+    // Update P2P status and emit events
+    this._setP2PStatus(true);
+
+    // Remove remote tracks
+    this._removeRemoteJVBTracks();
+
+    // Add remote tracks
+    this._addRemoteP2PTracks();
+
+    // Remove local tracks from JVB PC
+    // But only if it has started
+    if (this.jvbJingleSession) {
```

Comment:
1. It will extend the process
2. It will make the transition more noticeable (the black video flash) mostly because if the P2P is established first there will be no effect at all
3. I don't see too much improvements this would bring. There are not so many checks around that and I guess those will still be required to some extent